test/e2e_env/kubernetes/gateway/gateway.go

```diff
@@ -578,6 +579,121 @@ spec:
+		httpRoute := fmt.Sprintf(`
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: http-route-1
+  namespace: %s
+  labels:
+    kuma.io/mesh: %s
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: simple-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: "/timeout"
+      default:
+        backendRefs:
+        - kind: MeshService
+          name: "echo-server_simple-gateway_svc_80"`, Config.KumaNamespace, meshName)
+		httpRoute2 := fmt.Sprintf(`
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: http-route-2
+  namespace: %s
+  labels:
+    kuma.io/mesh: %s
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: simple-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: "/no-timeout"
+      default:
+        backendRefs:
+        - kind: MeshService
+          name: "echo-server_simple-gateway_svc_80"`, Config.KumaNamespace, meshName)
+		mt := fmt.Sprintf(`
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: mesh-timeout-1
+  namespace: %s
+  labels:
+    kuma.io/mesh: %s
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: http-route-1
+  to:
+    - targetRef:
+        kind: Mesh
+      default:
+        http:
+          requestTimeout: 2s
+          streamIdleTimeout: 10s`, Config.KumaNamespace, meshName)
+
+		BeforeAll(func() {
+			err := NewClusterSetup().
+				Install(YamlK8s(httpRoute, httpRoute2)).
+				Install(YamlK8s(mt)).
+				Setup(kubernetes.Cluster)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		AfterAll(func() {
+			err := NewClusterSetup().
+				Install(DeleteYamlK8s(mt)).
+				Install(DeleteYamlK8s(httpRoute, httpRoute2)).
+				Setup(kubernetes.Cluster)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should timeout only on one route", func() {
+			// timeout on specific route
+			Eventually(func(g Gomega) {
+				response, err := client.CollectFailure(
+					kubernetes.Cluster, "demo-client", "http://simple-gateway.simple-gateway:8080/timeout",
+					client.FromKubernetesPod(clientNamespace, "demo-client"),
+					client.WithHeader("host", "example.kuma.io"),
+					client.WithHeader("x-set-response-delay-ms", "3000"),
+					client.WithMaxTime(10),
+				)
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(response.ResponseCode).To(Equal(504))
+			}, "1m", "1s", MustPassRepeatedly(5)).Should(Succeed())
+
+			// no timeout on another route
+			Eventually(func(g Gomega) {
+				start := time.Now()
+				_, err := client.CollectEchoResponse(
+					kubernetes.Cluster, "demo-client", "http://simple-gateway.simple-gateway:8080/no-timeout",
+					client.FromKubernetesPod(clientNamespace, "demo-client"),
+					client.WithHeader("host", "example.kuma.io"),
+					client.WithHeader("x-set-response-delay-ms", "3000"),
+					client.WithMaxTime(10), // we don't want 'curl' to return early
```

Comment:
why would curl return early? 🤔

Reply:
Setting explicitly ensures that it won't finish in less than 10 seconds and it should finish in > 3.

Reply:
But why is it important that we won't finish in less than 10 seconds? can you explain this and add a comment?

Reply:
The default timeout is 5 seconds, and since we timeout in 2 seconds, the default is sufficient. There was no need for a timeout of 10 seconds for curl; it must be higher than the Mesh timeout to prevent curl from failing but still timing out due to the Mesh timeout.